### ISSUE TITLE
Proper Ad support

### ISSUE THREAD

**Issue by JasonLG1979** [pithos/pithos#292]
Reading up on the docs and poking though pandora.py I believe I've come up with proper ad support for Pithos. Turns out we were skipping ads all alone. Every playlist for free users contains a "song" that just contains an adToken. We were skipping it by checking for 'songName' in the playlist values.

Proper ad support has it's pluses and minuses.

On the plus side you can actually see ads in the playlist before they play, we would be supporting Pandora's revenue stream(no Pandora, no Pithos) since the ads can be "registered" so Pandora knows you heard them and hopefully as a side effect of that we could get less ads masquerading as songs.

On the minus side it would probably mean more ads for free users(Pandora One users never get ads). As I said before so far in my tests there's been an ad in just about every playlist.(so every 5th track), although there's still nothing to stop people from skipping them if they want.

Here's a screenshot of an upcoming ad in the playlist of my testing free account.
 
![screenshot from 2016-04-17 05-28-21](https://cloud.githubusercontent.com/assets/6667703/14586485/4ddc1c0a-045f-11e6-9d2b-ff9805fab680.png)

**Comment by JasonLG1979** [comment]
Update:
It seems that there aren't as many ads as I thought.(1 per playlist) Pandora does send an adToken for every playlist but most of the time the metadata returned only contains an empty list of adTrackingTokens. I take that to mean there is not ad at that time or Pandora doesn't want to place an ad in the playlist? In my tests actual ads appear no more frequently then in any other client(maybe even slightly less) I'm going to do some more testing and fiddling with optional api call args and I'll throw some code up in a branch probably in a couple days or so.

**Comment by MadcapJake** [comment]
I have never gotten an ad with Pithos and I use it all day.  On Pandora's web client, I get ads all the time (perhaps making up for all my Pithos time).

**Comment by TingPing** [comment]
Users would certainly be displeased by getting *a lot* more ads. :P

**Comment by JasonLG1979** [comment]
> I have never gotten an ad with Pithos and I use it all day.

That is because technically we skip ads. In pandora.py, notice the comment:

```
    def get_playlist(self):
        logging.info("pandora: Get Playlist")
        playlist = self.pandora.json_call('station.getPlaylist', {'stationToken': self.idToken, 'includeTrackLength': True}, https=True)
        songs = []
        for i in playlist['items']:
            if 'songName' in i: # check for ads
                songs.append(Song(self.pandora, i))
        return songs
```
I don't think the purpose was to not play ads per-say I think it was to avoid exceptions/errors, as " ad songs" only contain an adToken and no other metadata. I'm going to guess that no one knew what to do with said adToken. Otherwise as far as ads in Pithos very rarely an ad will sneak into a playlist pretending to be and otherwise be distinguishable from a song except for the audio ofc.    

> On Pandora's web client, I get ads all the time (perhaps making up for all my Pithos time).

I'm not sure how that works exactly.

**Comment by JasonLG1979** [comment]
> Users would certainly be displeased by getting a lot more ads. :P

At this point an ad every 5 days would be "a lot more."

I don't think anyone really "wants" ads but like I said earlier, without ads there is no Pandora and without Pandora there is no Pithos. So I think for that reason and ethically we should try to do the right thing. Some people I'm sure will confuse free with freeloader and bock at the ads but my answer to that is if you don't want ads spend the $5 a month on a Pandora One subscription.

**Comment by MadcapJake** [comment]
I think it's ethically important that Pithos provide ads. I want to support what Pandora is doing but I also want to use nice Gnome apps whenever possible.  I'd like to do both at the same time. :smile:

**Comment by JasonLG1979** [comment]
After some more totally unscientific testing I'm getting an average, more or less 2 30 sec ads per hour or so or about 2 ads per 20 songs in my Quickmix. From what I understand ad frequency may be different depending on the station. Stations with newer and/or more popular songs may contain more ads then stations with older/less popular songs.(royalty price differences?)

**Comment by Maggew** [comment]
I think this is a dumbass idea - please do not integrate. 

In fact, if Pandora really cared about providing a maximum quality experience for their targeted audience, they would have created a Linux desktop app from day one. Also, if Pandora crashes and burns because ads are being skipped then they have a weak business model and rightfully so, deserve to crash and burn.

You know what's ethical? Worrying about problems that actually matter in this world. Just listen to that Steve Miller Band song "Fly Like An Eagle".

**Comment by JasonLG1979** [comment]
It would be awesome if this could be kept constructive...

But like I said...
> I'm sure [some] will confuse free with freeloader and bock at the ads but my answer to that is if you don't want ads spend the $5 a month on a Pandora One subscription.

**Comment by Maggew** [comment]
It's not about the money, it's about principle. 

Pandora has the power to create an awesome multi-platform desktop app. Instead, some (random cool) dude invested tons and tons of hours to do their job. Meanwhile...some dude wants to funnel money into Pandora and just skip that cool, random dude.

**Comment by JasonLG1979** [comment]
> It's not about the money, it's about principle.
> 
> Pandora has the power to create an awesome multi-platform desktop app. Instead, some (random cool) dude invested tons and tons of hours to do their job. Meanwhile...some dude wants to funnel money into Pandora and just skip that cool, random dude.

I agree it is about principle. Pandora does have the power to create an awesome multi-platform desktop app but they have chosen not to, If I had to guess because they did the math and figured it wasn't worth it financially to them. They exist to make money. That's their job. That's their driving  principle. Not giving you free stuff does not violate that principle. In fact giving you anything for free would be the violation. You or I are not entitled to a Pandora desktop app, the owe us nothing. You seem to confused in thinking that Pandora is free at all. They provide their service to you at no financial cost to you in exchange for putting up with an ad from time to time. 

As far as the "dudes" are concerned there are many, many "dudes" that have contributed to Pithos, you'll have to be more specific..

**Comment by TingPing** [comment]
> Meanwhile...some dude wants to funnel money into Pandora and just skipover that cool, random dude.

That is really an over-arching problem with all foss applications and funding.

**Comment by JasonLG1979** [comment]
> That is really an over-arching problem with all foss applications and funding.

I agree. But this isn't about lining Pandora's pockets while skipping over yours(or anyone else that contributes to Pithos) My only interest is no sabotaging Pandora's business model. We can talk all day about what Pandora should or shouldn't do but the fact still remains by skipping ads on purpose is in a way stealing the service(you're consuming it without "paying" for it by putting up with ads) it's a little hypocritical to use their perceived lack of "principles"  as a justification for stealing something totally nonessential. At that point you should maybe take a look at your own principles.

**Comment by JasonLG1979** [comment]
@TingPing What are your thoughts on this? I have perfectly working code for this but really you're the gate keeper so if you say it's a no go you can close this issue otherwise I can put in a PR to enable it?

**Comment by JasonLG1979** [comment]
A compromise would to enable ad support but not do anything to prevent users from skipping them(unlike the web client)?

**Comment by heniokalinowski** [comment]
Quoting Richard Stallman: "The answer is simple - the answer is freedom".

Problem like this is very common in software engineering, and best solution is:

https://en.wikipedia.org/wiki/Strategy_pattern

In other words, create two strategies: one that keeps skipping ads and one that plays them, plus check box.

Don't decide for people, give them choice to follow their morality, religion and political views, this is what free as in freedom means.

You are free to make playing ads by default on.
If You will try to force people to listen to ads, someone will for sure fork this project and maintain version just same, but without ads - do we really need it? No.

**Comment by TingPing** [comment]
I'm sorry but you can't be quoting RMS on freedom when referring to a proprietary web service.

**Comment by heniokalinowski** [comment]
I am not referring to proprietary web service.
I am referring to this open source project. that "is not affiliated with or endorsed by Pandora Media, Inc. " (as stated on https://pithos.github.io).

This solution is only way, to satisfy everyone, because everyone will be able to choose what to do, instead of being forced by choses made by others.

**Comment by JasonLG1979** [comment]
There's already a ad free option it's called a Pandora One account...

**Comment by heniokalinowski** [comment]
Pity that I cannot buy it because I am not in USA, and even if I would be, then I can't pay with bitcoins, because they want whole credit card (that I don't even have, and for sure won't create any) and all my personal data, that in case of leak no one will even say "we are sorry".

Trust me, it's not about few dollars, if it would solve this trouble no one would ever waste time creating this issue.

**Comment by Maggew** [comment]
Goto bed @JasonLG1979 - it's time for you to sleep...

Thanks for that brilliant reply @heniokalinowski - you are a superhero...

**Comment by JasonLG1979** [comment]
@Maggew, so far your contributions to this discussion have basically amounted to "That sucks, you're stupid, I want free stuff, shut up!!!" Unless you can be civil and have something intelligent/constructive to say keep it to yourself and let the grownups talk...

@heniokalinowski There are no easy answers there...
As far as not being able *legally* access Pandora in your area, I'd blame the RIAA and other trade groups like it. I'm sure Pandora would love to take your money but I'm guessing they can't get a licence for the music in your area. Pithos has Proxy support to help side step that.

As far as your personal anonymity security is concerned, I guess in a perfect world be could have 100% anonymous and secure access to any service we desire. But for many reasons unfortunately we don't at this time.

I'd say as nicely as possible that you really don't have anything to complain about. You're not suppose to get Pandora at all because of the regional restrictions. You really can't complain about a few ads.

**Comment by heniokalinowski** [comment]
I am not saying "remove ads" I say "give people choice". I really do not want to get into religion, political visions and other deep stuff, but You are basically trying to enforce Your views on others - history proves, that it never ended well.

Pandora right now does not have desktop client (especially for Linux) so their expected income from this platform equals zero. We are all here, just because such official client does not exist. And as You see, people are overcoming this issue, by creating their own client. If You will make next restriction, like forcing ads, then someone will overcome it with fork. As simple as that.

You are basically saying: "people must learn our peaceful ways.... by force!"

Closed source is about money, open source is about people.

**Comment by JasonLG1979** [comment]
You're the one that keeps bringing up politics and religion for some reason?.. It not about my views it's about trying to do things properly from a technical stand point(use the the API correctly as we can) and a ethical point of view(very basically skipping ads is stealing you can paint it how you like and call it what you want but it's taking the service without paying for it.)

**Comment by heniokalinowski** [comment]
Very basically, I as root/administrator decide, if my computer should play or not play byte stream.
Give people freedom of choice.

**Comment by JasonLG1979** [comment]
My PR doesn't stop you from skipping ads.

**Comment by heniokalinowski** [comment]
Ok, thats about me, and what about others?
Why do You not want to give people freedom of choice?

**Comment by JasonLG1979** [comment]
What about you? In my PR ads are pretty much treated just like a song(except you can't rate them and such) as such you can skip them just like any other song...

**Comment by heniokalinowski** [comment]
To be honest, I like certain ads, especially about IT and that are funny, however listening for 10th time to buy burger from company that does not even sell them in my country isn't best way to attract customers, and I will tell You even more - I have never seen this company, but I already dislike them for this.

I and those advertisers are mutually useless for ourselves.
They cannot attract me, because I won't fly to other part of world just to buy burger (so I am useless for them), and they take my time (so are useless for me).

Please note, that such burger company pays per ad, not per attracted customer, so basically are wasting money to advertise to me without (positive) effect.

I would be glad to hear ads about new super cheap VPS instead, but no one advertises it through Pandora as this service is not (only) targeting IT guys. So to save both my time and advertisers money it would be wise to do not even start playing this ad.

**Comment by JasonLG1979** [comment]
Well you can't complain about ads not being relevant to you if your using a proxy to listen in a restricted area. And the advertisers do pay by the ad I would guess. The API has a means to register ads so that Pandora know you heard them.

**Comment by heniokalinowski** [comment]
It is not my fault nor choice, that I have been born outside of USA.
You cannot blame people for following other than Yours morality.
Give people freedom of choice, and if Your morality is right, then they will follow it.

**Comment by JasonLG1979** [comment]
I know of very few places on earth were stealing is not frowned upon...

**Comment by TingPing** [comment]
Ok, well this discussion isn't going anywhere so I'll just lock it for now.
